p2p/udp.go

```diff
@@ -0,0 +1,226 @@
+package p2p
+
+import (
+	"errors"
+	"fmt"
+	"net"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/spacemeshos/go-spacemesh/log"
+	inet "github.com/spacemeshos/go-spacemesh/p2p/net"
+	"github.com/spacemeshos/go-spacemesh/p2p/node"
+	"github.com/spacemeshos/go-spacemesh/p2p/p2pcrypto"
+	"github.com/spacemeshos/go-spacemesh/p2p/pb"
+	"github.com/spacemeshos/go-spacemesh/p2p/service"
+)
+
+const maxMessageSize = 2048
+
+// todo : calculate real udp max message size
+
+// Lookuper is a service used to lookup for nodes we know already
+type Lookuper interface {
+	InternalLookup(key node.DhtID) []node.Node
+}
+
+type udpNetwork interface {
+	Start() error
+	Shutdown()
+
+	IncomingMessages() chan inet.UDPMessageEvent
+	Send(to node.Node, data []byte) error
+}
+
+// UDPMux is a server for receiving and sending udp messages. through protocols.
+type UDPMux struct {
+	local    *node.LocalNode
+	lookuper Lookuper
+	network  udpNetwork
+	messages map[string]chan service.DirectMessage
+	shutdown chan struct{}
+	logger   log.Log
+}
+
+// NewUDPMux creates a new udp protocol server
+func NewUDPMux(localNode *node.LocalNode, lookuper Lookuper, udpNet udpNetwork, logger log.Log) *UDPMux {
+	return &UDPMux{
+		localNode,
+		lookuper,
+		udpNet,
+		make(map[string]chan service.DirectMessage),
+		make(chan struct{}, 1),
+		logger,
+	}
+}
+
+// Start starts the UDPMux
+func (mux *UDPMux) Start() error {
+	err := mux.network.Start()
+	if err != nil {
+		return err
+	}
+	go mux.listenToNetworkMessage()
+	return nil
+}
+
+// Shutdown closes the server
+func (mux *UDPMux) Shutdown() {
+	close(mux.shutdown)
+	mux.network.Shutdown()
+}
+
+func (mux *UDPMux) listenToNetworkMessage() {
+	msgChan := mux.network.IncomingMessages()
+	for {
+		select {
+		case msg, ok := <-msgChan:
+			if !ok {
+				// closed
+				return
+			}
+			go func(event inet.UDPMessageEvent) {
+				err := mux.processUDPMessage(event.FromAddr, event.Message)
+				if err != nil {
+					mux.logger.Error("Error handing network message err=%v", err)
+					// todo: blacklist ?
+				}
+			}(msg)
+		case <-mux.shutdown:
+			return
+		}
+	}
+}
+
+// Note: for now udp is only direct.
+
+// todo: no need to return chan, but stay consistent with api
+// RegisterDirectProtocolWithChannel registers a protocol on a channel
+func (mux *UDPMux) RegisterDirectProtocolWithChannel(name string, c chan service.DirectMessage) chan service.DirectMessage {
+	mux.messages[name] = c
```

Comment:
If this method can be called at any time by any routine `messages` needs to be thread safe

Reply:
this method should be called only right after initialization, before `Start`. means `ProccessDirectProtocolMessage` - the routine that reads from the map should never be called concurrently with it.

Reply:
Add comment about it